feat(launcher): add information about which user launched a study

MartinBelthle opened the pull request:
The front-end part of the issue hasn't been coded yet.
But it can still be merged as it should be use as a basis to solve another issue : https://github.com/AntaresSimulatorTeam/AntaREST/issues/1732

⚠️ WARNING: this feature introduce a database breaking change (database migration).

MartinBelthle committed:
feat(launcher): add information about which user launched a study

ghost commented:
### Features

*  **launcher:** add information about which user launched a study ([c7d829f](https://github.com/AntaresSimulatorTeam/AntaREST/commit/c7d829f904bdae1004d8f76d7a47335b7f524e66))
*  **launcher:** add alembic migration script that works with sqlite ([9171d9d](https://github.com/AntaresSimulatorTeam/AntaREST/commit/9171d9d90f9458620061356bcbfb88c22582a1b6))
*  **alembic:** rename migration script with owne_id ([1eb7db5](https://github.com/AntaresSimulatorTeam/AntaREST/commit/1eb7db58888d65c21dea4947c07d7b8fe9177f24))
*  **job-status:** add the `owner_id` attribute to the `JobStatus` class ([a47574a](https://github.com/AntaresSimulatorTeam/AntaREST/commit/a47574ac07854576833544ff127a2cde0b0ef882))
*  **scripts:** update the alembic downgrade version ([a33dfbd](https://github.com/AntaresSimulatorTeam/AntaREST/commit/a33dfbd62712efd61b0226acff5124edef1f42d6))


### Bug Fixes

*  **launcher:** change default value to unknown user ([4e81e5f](https://github.com/AntaresSimulatorTeam/AntaREST/commit/4e81e5f991c7b703f2b9d504888a6be062dd220a))
*  **database:** use FK on primary key to work with Postgre ([f3b389f](https://github.com/AntaresSimulatorTeam/AntaREST/commit/f3b389ff2c8c2d128375de9365b41f240a984ad8))


### Tests

*  **sqlite:** use `PRAGMA foreign_keys = ON` to check foreign keys constraints ([7ad8254](https://github.com/AntaresSimulatorTeam/AntaREST/commit/7ad825405bbeb595bf5d9650ed5422cb5bf3c62d))
*  **launcher-model:** improve test coverage for Antares launcher model classes ([d41fcff](https://github.com/AntaresSimulatorTeam/AntaREST/commit/d41fcff1eec3ffdd3502a19ba57d5c6d4e9a5886))


### Chore

*  **scripts:** improve the bash scripts used for developers ([2cfabe4](https://github.com/AntaresSimulatorTeam/AntaREST/commit/2cfabe4c7f638fa7eb66b7d7812c578ca7d656a6))


### Documentation

*  **database:** update database management documentation ([87ad03d](https://github.com/AntaresSimulatorTeam/AntaREST/commit/87ad03d8aa55d4bff46b908c5fd348752f17609d))


### Contributors

<a href="https://github.com/MartinBelthle">MartinBelthle</a>, <a href="https://github.com/laurent-laporte-pro">laurent-laporte-pro</a>

<details>
<summary>Commit-Lint commands</summary>
<br />

You can trigger Commit-Lint actions by commenting on this PR:
- `@Commit-Lint merge patch` will merge dependabot PR on "patch" versions (X.X.Y - Y change)
- `@Commit-Lint merge minor` will merge dependabot PR on "minor" versions (X.Y.Y - Y change)
- `@Commit-Lint merge major` will merge dependabot PR on "major" versions (Y.Y.Y - Y change)
- `@Commit-Lint merge disable` will desactivate merge dependabot PR
- `@Commit-Lint review` will approve dependabot PR
- `@Commit-Lint stop review` will stop approve dependabot PR



</details>

MartinBelthle committed:
fix(launcher): change default value to unknown user

MartinBelthle committed:
checkstyle with black

MartinBelthle committed:
feat(launcher): add alembic migration script that works with sqlite

MartinBelthle committed:
fix(database): use FK on primary key to work with Postgre

MartinBelthle committed:
feat(alembic): rename migration script with owne_id

MartinBelthle commented:
From what I've tested, the alembic script works for Sqlite and for a postgreSQL DB inside a docker container.

laurent-laporte-pro commented on antarest/launcher/model.py:
```diff
@@ -121,6 +123,7 @@ class JobResult(DTO, Base):  # type: ignore
     exit_code = Column(Integer)
     solver_stats = Column(String(), nullable=True)
     logs = relationship(JobLog, uselist=True, cascade="all, delete, delete-orphan")
+    owner_id = Column(Integer(), ForeignKey(Identity.id), default=0)
```
En fait, lorsqu'on a une colonne qui référence une clé étrangère de manière optionnelle, on ne peut pas mettre de valeur par défaut autre que `None` (soit `NULL` en SQL). Le mieux ici est d'ajouter le paramètre `nullable=True`. Car, si tu mets `0`, cela signifie que tu fais référence à un enregistrement dont l'ID est `0`.

Cela a donc un impact dans la définition du `JobResultDTO`, il faut là aussi autoriser la valeur `None` (que je déteste).

laurent-laporte-pro replied:
Les tests unitaires ne mettent malheureusement pas en évidence ce problème car les contraintes de clés étrangères ne sont pas vérifiées par défaut avec SQLite. Pour ce faire, il faudrait les activer avec `PRAGMA foreign_keys = ON`.

L'activation est cependant possible dans la fixture `db_engine_fixture` :

```python
@pytest.fixture(name="db_engine")
def db_engine_fixture() -> Generator[Engine, None, None]:
    engine = create_engine("sqlite:///:memory:")
    engine.execute("PRAGMA foreign_keys = ON")  # <- ICI
    Base.metadata.create_all(engine)
    yield engine
    engine.dispose()
```

laurent-laporte-pro replied:
Par la suite, si l'on supprime un enregistrement de type `Identity` qui est lié à un (ou plusieurs) `JobResult`, on souhaite mettre le `owner_id` à `None`.
Pour ce faire, il convient d'avoir la déclaration suivante :

```python
owner_id = Column(Integer(), ForeignKey(Identity.id, ondelete="SET NULL"), nullable=True)
```

Le paramètre `ondelete="SET NULL"` indique à la base de données de mettre à `NULL` toutes les clés étrangères lorsqu'il y a suppression de l'enregistrement référencé.
Cette information doit également figurer dans le script de migration.

laurent-laporte-pro commented on antarest/launcher/service.py:
```diff
@@ -224,12 +224,16 @@ def run_study(
             study=study_info,
             permission_type=StudyPermissionType.RUN,
         )
+        owner_id = 0
```
Voici le texte corrigé :

La valeur par défaut du `owner_id` doit être `None`, mais comme les IDs de `impersonator` ou de `user` peuvent être nuls, il vaut mieux les convertir en `None` lorsqu'on a `0` au moment de la création de l'objet `JobResult` :

```python
job_status = JobResult(
    id=job_uuid,
    study_id=study_uuid,
    job_status=JobStatus.PENDING,
    launcher=launcher,
    launcher_params=launcher_parameters.json() if launcher_parameters else None,
    owner_id=(None if owner_id == 0 else owner_id),
)
```

laurent-laporte-pro commented on antarest/launcher/service.py:
```diff
@@ -267,11 +271,15 @@ def kill_job(self, job_id: str, params: RequestParameters) -> JobResult:
 
         self.launchers[launcher].kill_job(job_id=job_id)
 
+        owner_id = 0
```
Même remarque.

laurent-laporte-pro committed:
test(sqlite): use `PRAGMA foreign_keys = ON` to check foreign keys constraints

laurent-laporte-pro committed:
feat(job-status): add the `owner_id` attribute to the `JobStatus` class

- Correct issue with the foreign key: add `ondelete="SET NULL"`
- Add documentation to the `JobResultDTO` class
- Remove comparison operators `__eq__` (and `__ne__`)

laurent-laporte-pro reviewed (dismissed):
Pourrais-tu re-faire des tests avec PostgreSQL ?

laurent-laporte-pro committed:
chore(scripts): improve the bash scripts used for developers

laurent-laporte-pro committed:
feat(scripts): update the alembic downgrade version

laurent-laporte-pro committed:
docs(database): update database management documentation

Improve documentation regarding database management for different environments using SQLAlchemy and Alembic. Clarify the use of PostgreSQL for production and SQLite for local applications. Also, provide details about how to update the schema and manage migrations in the application.

laurent-laporte-pro commented:
See [Alembic migration documentation](https://antares-web.readthedocs.io/en/latest/architecture/1-database/)

laurent-laporte-pro committed:
test(launcher-model): improve test coverage for Antares launcher model classes
